### PR TITLE
[FLINK-32079][checkpoint] Support to read/write checkpoint metadata of merged files

### PR DESCRIPTION
**Pull request by masteryhx**
## What is the purpose of the change

Support to read/write checkpoint metadata of merged files

## Brief change log

  - serialize and deserialize SegmentFileStateHandle in MetadataV2V3SerializerBase
  - hotfix error equals for SegmentFileStateHandle

## Verifying this change

This change added tests and can be verified as follows:

  - Add _testSerializeSegmentStateHandle_

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), **Checkpointing**, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by masteryhx** [comment]
@Zakelly @fredia Could you take a look ?

**Comment by Zakelly** [comment]
> @Zakelly @fredia Could you take a look ?

@masteryhx 
How about waiting for the FLINK-34668, where some new state handles will be introduced.

**Comment by masteryhx** [comment]
> > @Zakelly @fredia Could you take a look ?
> 
> @masteryhx How about waiting for the [FLINK-34668](https://issues.apache.org/jira/browse/FLINK-34668), where some new state handles will be introduced.

Thanks for the suggestion.
I just rework it after rebasing [FLINK-34668](https://issues.apache.org/jira/browse/FLINK-34668) .

**Review by Zakelly — COMMENTED** [review]
Thanks for this PR! Minor question:

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV2V3SerializerBase.java [hunk 1]
```diff
@@ -677,6 +734,18 @@ static void serializeStreamStateHandle(StreamStateHandle stateHandle, DataOutput
             RelativeFileStateHandle relativeFileStateHandle = (RelativeFileStateHandle) stateHandle;
             dos.writeUTF(relativeFileStateHandle.getRelativePath());
             dos.writeLong(relativeFileStateHandle.getStateSize());
+        } else if (stateHandle instanceof SegmentFileStateHandle) {
+            if (stateHandle instanceof EmptySegmentFileStateHandle) {
+                dos.writeByte(EMPTY_SEGMENT_FILE_HANDLE);
```
Comment (Zakelly) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV2V3SerializerBase.java:739]: IIUC, there is not a chance that `EmptySegmentFileStateHandle` is serialized here?

Comment (masteryhx) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV2V3SerializerBase.java:739]: This `serializeStreamStateHandle` is used when serialize `EmptyFileMergingOperatorStreamStateHandle` which holds this as its delegated empty stream state handle.
So it's still needed, right ?

Comment (Zakelly) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV2V3SerializerBase.java:739]: Ah.... I missed the `serializeStreamStateHandle(stateHandle.getDelegateStateHandle(), dos);` part

**Review by Zakelly — APPROVED** [review]
Thanks for the update! LGTM+1

**Pull request closed without merging**